dinput_hook/renderer_utils.cpp

```diff
@@ -0,0 +1,428 @@
+#include "renderer_utils.h"
+
+#include <glad/glad.h>
+#include <GLFW/glfw3.h>
+#include <string>
+#include <windows.h>
+#include <cmath>
+
+#include "globals.h"
+#include "types.h"
+
+extern "C" {
+#include <Primitives/rdMatrix.h>
+}
+
+extern "C" FILE *hook_log;
+
+rdMatrix44 g_projectionMatrix;
+rdMatrix44 g_modelViewMatrix;
+bool g_useAlphaMask;
+bool g_useFog;
+float g_fogColor[4];
+float g_fogStart;
+float g_fogEnd;
+
+extern "C" __declspec(dllexport) void renderer_setOrtho(rdMatrix44 *m, float left, float right,
+                                                        float bottom, float top, float nearVal,
+                                                        float farVal) {
+    float tx = -(right + left) / (right - left);
+    float ty = -(top + bottom) / (top - bottom);
+    float tz = -(farVal + nearVal) / (farVal - nearVal);
+
+    memcpy(m,
+           (float[16]){2.0f / (right - left), 0, 0, 0, 0, 2 / (top - bottom), 0, 0, 0, 0,
+                       -2 / (farVal - nearVal), 0, tx, ty, tz, 1},
+           sizeof(*m));
+}
+
+extern "C" __declspec(dllexport) void renderer_setProjectionMatrix(rdMatrix44 *projectionMatrix) {
+    memcpy(&g_projectionMatrix, projectionMatrix, sizeof(g_projectionMatrix));
+}
+
+extern "C" __declspec(dllexport) void renderer_setModelViewMatrix(rdMatrix44 *modelViewMatrix) {
+    memcpy(&g_modelViewMatrix, modelViewMatrix, sizeof(g_modelViewMatrix));
+}
+
+/**
+ * TODO: Set an uniform that discard if useAlphaMask is enabled
+ */
+extern "C" __declspec(dllexport) void renderer_setAlphaMask(bool useAlphaMask) {
+    // Use discard a <= 0 instead
+    //      glEnable(GL_ALPHA_TEST);
+    //      glAlphaFunc(GL_GREATER, 0); // drawn if a > 0
+    // } else { // false
+    //      glDisable(GL_ALPHA_TEST);
+    g_useAlphaMask = useAlphaMask;
+}
+
+/**
+ * TODO: Set an uniform to do fog computation if enabled
+ */
+extern "C" __declspec(dllexport) void renderer_setFog(bool useFog) {
+    g_useFog = useFog;
+}
+
+extern "C" __declspec(dllexport) void renderer_setLinearFogParameters(float color[4], float start,
+                                                                      float end) {
+    memcpy(g_fogColor, color, sizeof(g_fogColor));
+    g_fogStart = start;
+    g_fogEnd = end;
+}
+
+GLuint compileProgram(GLsizei vertexCount, const GLchar **vertexShaderSource, GLsizei fragmentCount,
```

Comment:
compileProgram could add a prefix with the g_ uniforms for each shader

Reply:
I don't understand what you mean. Uniforms as an input parameter ? I'd rather keep it as simple as possible instead of doing yet another shader generator. Shader generators as code are horribly complex and hard to debug / inspect in my experience

Reply:
that was just an idea how to handle the global g_ variables in the file, i.e. adding all of them as uniforms to all shaders. but there are of course other ways to do it 👍